frameworks/hdfs/src/main/java/com/mesosphere/sdk/hdfs/scheduler/HdfsRecoveryPlanManager.java

```diff
@@ -0,0 +1,167 @@
+import com.mesosphere.sdk.scheduler.recovery.RecoveryType;
+import com.mesosphere.sdk.scheduler.recovery.constrain.LaunchConstrainer;
+import com.mesosphere.sdk.scheduler.recovery.monitor.FailureMonitor;
+import com.mesosphere.sdk.specification.PodInstance;
+import com.mesosphere.sdk.specification.ServiceSpec;
+import com.mesosphere.sdk.state.StateStore;
+import com.mesosphere.sdk.state.StateStoreUtils;
+import org.apache.mesos.Protos;
+
+import java.util.*;
+import java.util.function.Predicate;
+import java.util.stream.Collectors;
+
+/**
+ * The HdfsRecoveryPlanManager handles failure scenarios unique to HDFS.  It falls back to the default recovery behavior
+ * when appropriate.
+ */
+public class HdfsRecoveryPlanManager extends DefaultRecoveryPlanManager {
+    private final Plan replaceNN0Plan;
+    private final Plan replaceNN1Plan;
+    private static final String NN_PHASE_NAME = "permanent-nn-failure-recovery";
+    private Phase nameNodeRecoveryPhase = new DefaultPhase(
+            NN_PHASE_NAME,
+            Collections.emptyList(),
+            new SerialStrategy<>(),
+            Collections.emptyList());
+
+    public HdfsRecoveryPlanManager(
+            StateStore stateStore,
+            ConfigStore<ServiceSpec> configStore,
+            LaunchConstrainer launchConstrainer,
+            FailureMonitor failureMonitor,
+            Plan replaceNN0Plan,
+            Plan replaceNN1Plan) {
+        super(stateStore, configStore, launchConstrainer, failureMonitor);
+        this.replaceNN0Plan = replaceNN0Plan;
+        this.replaceNN1Plan = replaceNN1Plan;
+    }
+
+    @Override
+    public Plan getPlan() {
+        List<Phase> phases = new ArrayList<>();
+        List<String> nnStepNames = nameNodeRecoveryPhase.getChildren().stream()
+                .map(step -> step.getName())
+                .collect(Collectors.toList());
+
+        // Filter out steps handled by name node recovery from the default plan.
+        for (Phase phase : super.getPlan().getChildren()) {
+            List<Step> steps = phase.getChildren().stream()
+                    .filter(step -> !nnStepNames.contains(step.getName()))
+                    .collect(Collectors.toList());
+            phases.add(
+                    new DefaultPhase(
+                            phase.getName(),
+                            steps,
+                            new SerialStrategy<>(),
+                            Collections.emptyList()));
+        }
+
+        if (!nameNodeRecoveryPhase.getChildren().isEmpty()) {
+            phases.add(nameNodeRecoveryPhase);
+        }
+
+        return new DefaultPlan(
+                RECOVERY_ELEMENT_NAME,
+                phases,
+                new ParallelStrategy<>(),
+                Collections.emptyList());
+    }
+
+    @Override
+    public Collection<? extends Step> getCandidates(Collection<String> dirtyAssets) {
+        // Always allow the default recovery manager to update its plan.
```

Comment:
This comment feels like it was the result of some pain, if so, maybe clarify what the commented line prevents?